File: src/trailmate/ai/context.py
"""Token-budgeted context window manager for the AI service.

Tracks token spend across iterations and applies a tiered compaction
strategy when the conversation crosses a configured budget. Exposed
via three public methods:

- ``track_burn(usage)``       — accumulate token counters from an API response
- ``enforce_compaction(...)`` — return a compacted view of the history
                                (does NOT mutate the input)
- ``_summarize`` / ``_estimate_tokens`` — private helpers

Design notes:
- Compaction is *return-view only* — ``chat_history`` on the service stays
  intact. Each ``run()`` iteration recomputes compaction.
- Token estimation uses ``tiktoken`` for accurate counts on JSON-heavy
  tool messages.
- "Keep last N messages" is implemented as "keep last N *groups*" where
  a group is a non-tool message OR an assistant ``tool_calls`` message
  plus all its matching ``tool`` results. This guarantees no orphan tool
  messages, which the OpenAI API rejects.
- The summarization callable is injected at construction time so this
  class has no direct dependency on any LLM SDK.
"""

from __future__ import annotations

import json
from typing import Any, Callable

import tiktoken

from trailmate.logging_config import get_logger

logger = get_logger(__name__)


class ContextManager:
    """Owns token accounting + compaction for a chat-completion loop."""

    def __init__(
        self,
        max_context_tokens: int,
        summarize_fn: Callable[[str], str] | None = None,
    ) -> None:
        self.max_context_tokens = max_context_tokens
        self.metrics = {"input_tokens": 0, "output_tokens": 0, "total_tokens": 0}

        # Injected callable for step-3 summarization. If None, step 3 is
        # skipped and step 2 is the final fallback (safe for unit tests
        # that don't have a provider available).
        self._summarize_fn = summarize_fn

        try:
            self._encoding = tiktoken.encoding_for_model("gpt-4o")
        except KeyError:
            self._encoding = tiktoken.get_encoding("cl100k_base")

    def track_burn(self, usage: Any) -> None:
        """Add this API response's usage to running totals.

        Only prints when the cumulative input tokens exceed 60% of the
        budget — quiet during normal operation, visible when approaching
        the compaction threshold.
        """
        self.metrics["input_tokens"] += usage.prompt_tokens
        self.metrics["output_tokens"] += usage.completion_tokens
        self.metrics["total_tokens"] += usage.total_tokens

        pct = self.metrics["input_tokens"] / self.max_context_tokens
        if pct >= 0.60:
            logger.info(
                "token budget %.0f%% — in: %d / %d | out: %d",
                pct * 100,
                self.metrics["input_tokens"],
                self.max_context_tokens,
                self.metrics["output_tokens"],
            )

    def enforce_compaction(self, history: list, tool_use: Any) -> list:
        """Return a (possibly compacted) view of ``history`` under the budget.

        Does NOT mutate ``history``. Applies the three-step strategy in
        order, returning early as soon as the result fits the budget:

        1. Evict raw tool payloads (replace ``tool`` message content with a
           short placeholder; preserves message structure so tool_call/
           tool_response pairs still validate).
        2. Keep only the last 4 *groups*.
        3. Summarize everything between the system seed and the kept tail
           into a single synthetic system message.

        ``tool_use`` is unread but kept on the signature as a deliberate
        forward hook for future tool-aware compaction strategies.
        """
        del tool_use  # forward-compat hook; intentionally unused for now

        before_msgs = len(history)
        before_tokens = self._estimate_tokens(history)

        if before_tokens <= self.max_context_tokens:
            return history

        # ---- Step 1: replace tool result payloads with a placeholder. ----
        step1: list[dict] = []
        for msg in history:
            if msg.get("role") == "tool":
                step1.append({**msg, "content": "[tool result evicted]"})
            else:
                step1.append(msg)
        if self._estimate_tokens(step1) <= self.max_context_tokens:
            logger.info(
                "compaction step 1: %d msgs / %d tok → %d msgs / %d tok",
                before_msgs, before_tokens, len(step1), self._estimate_tokens(step1),
            )
            return step1

        # ---- Step 2: keep last 4 groups (atomic tool round-trips). ----
        system_msg: dict | None = None
        body_start = 0
        if step1 and step1[0].get("role") == "system":
            system_msg = step1[0]
            body_start = 1
        body = step1[body_start:]
        last_groups_flat = _keep_last_n_groups(body, n=4)
        step2 = ([system_msg] if system_msg else []) + last_groups_flat
        if self._estimate_tokens(step2) <= self.max_context_tokens:
            logger.info(
                "compaction step 2: %d msgs / %d tok → %d msgs / %d tok",
                before_msgs, before_tokens, len(step2), self._estimate_tokens(step2),
            )
            return step2

        # ---- Step 3: summarize the discarded middle into one message. ----
        kept_tail_len = len(last_groups_flat)
        middle = body[: len(body) - kept_tail_len]
        if not middle:
            logger.info(
                "compaction step 2 only (nothing to summarize): "
                "%d msgs / %d tok → %d msgs / %d tok",
                before_msgs, before_tokens, len(step2), self._estimate_tokens(step2),
            )
            return step2

        summary = self._summarize(middle)
        summary_msg = {
            "role": "system",
            "content": f"Previous context summary: {summary}",
        }
        result = ([system_msg] if system_msg else []) + [summary_msg] + last_groups_flat
        logger.info(
            "compaction step 3 (summarized): %d msgs / %d tok → %d msgs / %d tok",
            before_msgs, before_tokens, len(result), self._estimate_tokens(result),
        )
        return result

    def _summarize(self, history: list) -> str:
        """Compress ``history`` into a short summary via the injected callable."""
        rendered_lines: list[str] = []
        for msg in history:
            role = msg.get("role", "?")
            content = msg.get("content") or ""
            if msg.get("tool_calls"):
                calls = json.dumps(msg["tool_calls"], default=str)
                rendered_lines.append(f"{role}: [tool_calls={calls}] {content}")
            else:
                rendered_lines.append(f"{role}: {content}")
        history_text = "\n".join(rendered_lines)

        prompt = (
            "Summarize the following conversation history "
            "concisely, preserving key facts, decisions, and "
            "tool results:\n" + history_text
        )
        return self._summarize_fn(prompt)  # type: ignore[misc]

    def _estimate_tokens(self, context: list) -> int:
        """Sum tiktoken-measured token counts across every message."""
        total = 0
        for msg in context:
            try:
                serialized = json.dumps(msg, default=str)
            except (TypeError, ValueError):
                serialized = str(msg)
            total += len(self._encoding.encode(serialized))
        return total


def _keep_last_n_groups(messages: list, n: int) -> list:
    """Return the last ``n`` conversation groups as a flat list of messages.

    A "group" is either:
    - a single non-tool message (user, assistant text, etc.), OR
    - an assistant message with ``tool_calls`` plus every immediately
      following ``tool`` message (its results).

    This guarantees we never split a tool-call/tool-result pair across
    the eviction boundary, which would produce orphan tool messages and
    cause the OpenAI API to reject the next request.
    """
    groups: list[list[dict]] = []
    current: list[dict] = []

    for msg in messages:
        role = msg.get("role")
        if role == "tool":
            if not current:
                current = [msg]
            else:
                current.append(msg)
        else:
            if current:
                groups.append(current)
            current = [msg]
    if current:
        groups.append(current)

    kept_groups = groups[-n:]
    return [m for g in kept_groups for m in g]

File: web/server/agent/context.ts
// Token-budgeted context manager for the agent loop — TypeScript port of the
// Python prototype's `trailmate.ai.context.ContextManager`.
//
// Tracks token spend across iterations and applies a tiered compaction
// strategy when the conversation crosses a configured budget:
//
//   1. Evict raw tool payloads (replace `tool` message content with a short
//      placeholder; preserves message structure so tool_call/tool_response
//      pairs still validate).
//   2. Keep only the last N *groups* — a group is a non-tool message OR an
//      assistant `tool_calls` message plus all of its `tool` results, so a
//      tool round-trip is never split (orphan tool messages are rejected by
//      the OpenAI API).
//   3. Summarize everything between the system seed and the kept tail into
//      one synthetic system message, via the injected `summarizeFn`.
//
// Compaction is *return-view only* — the caller's message array is never
// mutated; each loop iteration recomputes the compacted view.

import { Tiktoken } from "js-tiktoken/lite";
import o200k_base from "js-tiktoken/ranks/o200k_base";
import { createLogger } from "../shared/logger";

const log = createLogger("agent.context");

// Any chat-completion message. Structural on purpose so the module doesn't
// depend on the OpenAI SDK's param types (mirrors the Python dict approach).
export interface ContextMessage {
  role: string;
  content?: unknown;
  tool_calls?: unknown[];
  tool_call_id?: string;
  [key: string]: unknown;
}

export interface TokenUsage {
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
}

// gpt-4o family uses the o200k_base encoding. One encoder per process — the
// rank table is large, so never construct this per request.
const encoding = new Tiktoken(o200k_base);

const KEEP_LAST_GROUPS = 4;

export class ContextManager {
  readonly maxContextTokens: number;
  readonly metrics = { input_tokens: 0, output_tokens: 0, total_tokens: 0 };
  private summarizeFn?: (prompt: string) => Promise<string>;

  constructor(opts: {
    maxContextTokens: number;
    summarizeFn?: (prompt: string) => Promise<string>;
  }) {
    this.maxContextTokens = opts.maxContextTokens;
    this.summarizeFn = opts.summarizeFn;
  }

  /** Accumulate an API response's usage. Logs only past 60% of budget. */
  trackBurn(usage: TokenUsage): void {
    this.metrics.input_tokens += usage.prompt_tokens;
    this.metrics.output_tokens += usage.completion_tokens;
    this.metrics.total_tokens += usage.total_tokens;

    const pct = this.metrics.input_tokens / this.maxContextTokens;
    if (pct >= 0.6) {
      log.info("token_budget", {
        pct: Math.round(pct * 100),
        inputTokens: this.metrics.input_tokens,
        budget: this.maxContextTokens,
        outputTokens: this.metrics.output_tokens,
      });
    }
  }

  /** Return a (possibly compacted) view of `history` under the budget. */
  async enforceCompaction(history: ContextMessage[]): Promise<ContextMessage[]> {
    const beforeMsgs = history.length;
    const beforeTokens = this.estimateTokens(history);

    if (beforeTokens <= this.maxContextTokens) return history;

    // ---- Step 1: replace tool result payloads with a placeholder. ----
    const step1 = history.map((msg) =>
      msg.role === "tool" ? { ...msg, content: "[tool result evicted]" } : msg,
    );
    if (this.estimateTokens(step1) <= this.maxContextTokens) {
      log.info("compaction", {
        tier: 1, beforeMsgs, beforeTokens, afterMsgs: step1.length, afterTokens: this.estimateTokens(step1),
      });
      return step1;
    }

    // ---- Step 2: keep last N groups (atomic tool round-trips). ----
    let systemMsg: ContextMessage | null = null;
    let bodyStart = 0;
    if (step1.length > 0 && step1[0].role === "system") {
      systemMsg = step1[0];
      bodyStart = 1;
    }
    const body = step1.slice(bodyStart);
    const lastGroupsFlat = keepLastNGroups(body, KEEP_LAST_GROUPS);
    const step2 = (systemMsg ? [systemMsg] : []).concat(lastGroupsFlat);
    if (this.estimateTokens(step2) <= this.maxContextTokens) {
      log.info("compaction", {
        tier: 2, beforeMsgs, beforeTokens, afterMsgs: step2.length, afterTokens: this.estimateTokens(step2),
      });
      return step2;
    }

    // ---- Step 3: summarize the discarded middle into one message. ----
    const middle = body.slice(0, body.length - lastGroupsFlat.length);
    if (middle.length === 0 || !this.summarizeFn) {
      // Nothing to summarize (or no LLM injected) — step 2 is the final fallback.
      log.info("compaction", {
        tier: 2, final: true, beforeMsgs, beforeTokens, afterMsgs: step2.length, afterTokens: this.estimateTokens(step2),
      });
      return step2;
    }

    const summary = await this.summarize(middle);
    const summaryMsg: ContextMessage = {
      role: "system",
      content: `Previous context summary: ${summary}`,
    };
    const result = (systemMsg ? [systemMsg] : [])
      .concat([summaryMsg])
      .concat(lastGroupsFlat);
    log.info("compaction", {
      tier: 3, beforeMsgs, beforeTokens, afterMsgs: result.length, afterTokens: this.estimateTokens(result),
    });
    return result;
  }

  /** Compress `history` into a short summary via the injected callable. */
  private async summarize(history: ContextMessage[]): Promise<string> {
    const renderedLines = history.map((msg) => {
      const role = msg.role ?? "?";
      const content = typeof msg.content === "string" ? msg.content : "";
      if (msg.tool_calls) {
        return `${role}: [tool_calls=${JSON.stringify(msg.tool_calls)}] ${content}`;
      }
      return `${role}: ${content}`;
    });

    const prompt =
      "Summarize the following conversation history concisely, preserving " +
      "key facts, decisions, and tool results:\n" +
      renderedLines.join("\n");
    return this.summarizeFn!(prompt);
  }

  /** Sum tiktoken-measured token counts across every message. */
  estimateTokens(context: ContextMessage[]): number {
    let total = 0;
    for (const msg of context) {
      let serialized: string;
      try {
        serialized = JSON.stringify(msg) ?? String(msg);
      } catch {
        serialized = String(msg);
      }
      total += encoding.encode(serialized).length;
    }
    return total;
  }
}

/**
 * Return the last `n` conversation groups as a flat message list. A "group"
 * is a single non-tool message, or an assistant `tool_calls` message plus
 * every immediately following `tool` result — kept atomic so eviction never
 * produces orphan tool messages.
 */
export function keepLastNGroups(messages: ContextMessage[], n: number): ContextMessage[] {
  const groups: ContextMessage[][] = [];
  let current: ContextMessage[] = [];

  for (const msg of messages) {
    if (msg.role === "tool") {
      if (current.length === 0) current = [msg];
      else current.push(msg);
    } else {
      if (current.length > 0) groups.push(current);
      current = [msg];
    }
  }
  if (current.length > 0) groups.push(current);

  return groups.slice(-n).flat();
}
